invenio_rdm_records/services/pids/manager.py

```diff
@@ -87,15 +78,56 @@ def _validate_identifiers(self, pids, errors):
         for scheme, id_ in identifiers:
             pids[scheme]["identifier"] = id_
 
+    def _validate_pids(self, pids, record, errors):
+        """Validate an iterator of PIDs.
+
+        This function assumes all pid schemes are supported by the system.
+        """
+        for scheme, pid in pids.items():
+            provider = self._get_provider(scheme, pid.get("provider"))
+            success, val_errors = provider.validate(record=record, **pid)
+            if not success:
+                errors.append({"field": f"pids.{scheme}", "messages": val_errors})
+
+    def _validate_record(self, pids, record, errors):
+        """Validate the record according to the PIDs' requirements.
+
+        Here we check if the record is compatible from the point of
+        view of the pids...
+            - ... it contains
+            - ... it would contain according to configured required pids
+
+        The responsibility lies with each provider since they are the ones
+        that know their criteria for a record that is complete enough to get
+        a PID.
+        """
+        schemes = set(pids.keys()) | set(self._required_schemes)
+        scheme_provider_names = [
+            # provider_name for an absent-but-required pid will be None
+            (scheme, pids.get(scheme, {}).get("provider"))
+            for scheme in schemes
+        ]
+        providers = [
+            self._get_provider(scheme, provider_name)
+            for scheme, provider_name in scheme_provider_names
+        ]
+
+        for provider in providers:
+            success, provider_errors = provider.validate_record(record)
+            if not success:
+                errors.extend(provider_errors)
+
     def validate(self, pids, record, errors=None, raise_errors=False):
         """Validate PIDs."""
+        # if errors is [] we have to use it
         errors = [] if errors is None else errors
         self._validate_pids_schemes(pids)
         self._validate_identifiers(pids, errors)
         self._validate_pids(pids, record, errors)
+        self._validate_record(pids, record, errors)
```

Comment:
Make the validate_record call private (_validate_record) and call it here with `record`.

I buy the argument that `record` is what is the closest to what is published and therefore makes sense to validate (although the reader may be surprised like I originally was). (I don't buy so much the argument about "another component, added in-between in the future, that could change the record metadata. You cannot predict that." - that could be said about anything in which case any validation we make would be useless because it could be undone by another component.)

Reply:
It is not a question of what arguments you buy. The validation makes sense on what will be used by the provider (the record). There is no guarantee that the `draft` will not be potentially changed by another component, before or after.